File: alarm_clock/src/networking.py
# networking.py
import machine, network
import config
import time

net_ssid = config.net_ssid
net_pwd = config.net_pwd


def do_connect():
    try:
        wlan = network.WLAN(network.WLAN.IF_STA)
        wlan.active(True)
        if not wlan.isconnected():
            print('connecting to network...')
            wlan.connect(net_ssid, net_pwd)
            # Setting an Timeout for 5 minutes (x) => start + x * 60
            start = time.time()
            end = start + 5 * 60

            while not wlan.isconnected():
                time.sleep_ms(100)
                if time.time() >= end:
                    raise Exception ("network connection can't be established")
                    break


        print('network config:', wlan.ipconfig('addr4'))
        connection_success = True
        return connection_success
    except RuntimeError:
        print("something went wrong with the network connection..")
    except OSError as ex:
        print(f"OSError: {ex}")
    except Exception as ex:
        print(f"Anderer Fehler ({type(ex)}): {ex}")

File: alarm_clock/src/Networking.cpp
#include "Networking.h"

#include <WiFi.h>

bool connectWiFi(const char* ssid, const char* password, uint32_t timeoutMs) {
    if (WiFi.status() == WL_CONNECTED) {
        return true;
    }

    Serial.println("connecting to network...");
    WiFi.mode(WIFI_STA);
    WiFi.begin(ssid, password);

    uint32_t start = millis();
    while (WiFi.status() != WL_CONNECTED) {
        delay(100);
        if (millis() - start >= timeoutMs) {
            Serial.println("network connection can't be established");
            return false;
        }
    }

    Serial.print("network config: ");
    Serial.println(WiFi.localIP());
    return true;
}
